src/relay/pass/call_graph.cc

```diff
@@ -0,0 +1,339 @@
+  //
+  // This is the major overhead for constructing a call graph because the
+  // post-order visitor will visit each AST node of the current function to
+  // figure out the dependencies between functions.
+  PostOrderVisit(func, [&](const Expr& expr) {
+    if (const GlobalVarNode* gvn = expr.as<GlobalVarNode>()) {
+      auto callee = GetRef<GlobalVar>(gvn);
+      cg_node->AddCalledGlobal(LookupGlobalVar(callee));
+    }
+  });
+}
+
+const CallGraphEntry* CallGraphNode::operator[](const GlobalVar& gv) const {
+  const_iterator cit = call_graph_.find(gv);
+  CHECK(cit != call_graph_.end())
+      << "GlobalVar " << gv->name_hint << " not found in the call graph!";
+  return cit->second.get();
+}
+
+CallGraphEntry* CallGraphNode::operator[](const GlobalVar& gv) {
+  const_iterator cit = call_graph_.find(gv);
+  CHECK(cit != call_graph_.end())
+      << "GlobalVar " << gv->name_hint << " not found in the call graph!";
+  return cit->second.get();
+}
+
+// Query the existence of a GlobalVar in the call graph. It creates an entry if
+// there is no such a node available.
+CallGraphEntry* CallGraphNode::LookupGlobalVar(const GlobalVar& gv) {
+  CHECK(gv.defined());
+
+  // This inserts an element to the call graph if it is not there yet.
+  auto& call_graph_node = call_graph_[gv];
+  if (call_graph_node) return call_graph_node.get();
+
+  CHECK(module->ContainGlobalVar(gv->name_hint))
+      << "GlobalVar " << gv->name_hint << " not found in the current ir module";
+
+  // Create the node for the inserted entry.
+  call_graph_node = std::unique_ptr<CallGraphEntry>(new CallGraphEntry(gv));
+  return call_graph_node.get();
+}
+
+void CallGraphNode::Print(std::ostream& os) const {
+  // Print the call graph in the topological order.
+  std::vector<CallGraphEntry*> nodes = TopologicalOrder();
+  for (const auto* cgn : nodes) {
+    cgn->Print(os);
+  }
+}
+
+GlobalVar CallGraphNode::RemoveGlobalVarFromModule(CallGraphEntry* cg_node,
+                                                   bool update_call_graph) {
+  CHECK(cg_node->empty() || (cg_node->IsRecursive() && cg_node->size() == 1))
+      << "Cannot remove global var " << cg_node->GetNameHint()
+      << " from call graph, because it still calls "
+      << cg_node->size() << " other global functions";
+
+  if (update_call_graph) {
+    // Update the call graph by removing all edges that point to the node
+    // `cg_node`.
+    for (auto& it : *this) {
+      it.second->RemoveAllCallTo(cg_node);
+    }
+  }
+  GlobalVar gv = cg_node->GetGlobalVar();
+  call_graph_.erase(gv);
+  // Update the IR module.
+  module->Remove(gv);
+  return gv;
+}
+
+std::vector<CallGraphEntry*> CallGraphNode::GetEntryGlobals() const {
+  std::vector<CallGraphEntry*> ret;
+  // An entry function in Relay is a function that never called by other
+  // functions or only called by itself.
+  for (const auto& it : *this) {
+    if (it.second->GetRefCount() == 0 || it.second->IsRecursiveEntry()) {
```

Comment:
Might be dumb question -> But if `mod["g0"]` calls `mod["g1"]` and `mod["g1"]` is recursive, should `g1` be an Entry Global?>

Reply:
g0 should be, but not g1 in this case.